
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Using CORESH locally &#8212; CORESH Documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'pages/coresh-local';</script>
    <link rel="icon" href="../_static/coresh_o.png"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="References" href="references.html" />
    <link rel="prev" title="CORESH Hallmark Gene Set Results" href="hallmarks.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="input.html">
  
  
  
  
  
  
    <p class="title logo__title">CORESH Documentation</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="input.html">
                    Search Input
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="output.html">Search Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="study-cases.html">Study Cases</a></li>
<li class="toctree-l1"><a class="reference internal" href="hallmarks.html">CORESH Hallmark Gene Set Results</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Using CORESH locally</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
<li class="toctree-l1"><a class="reference internal" href="contacts.html">Contact Us</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/vdsukhov/coresh_frontend" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/pages/coresh-local.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Using CORESH locally</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#installing-dependencies">Installing dependencies</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#getting-the-data">Getting the data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#running-coresh-analysis-locally">Running CORESH analysis locally</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-info">Session info</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="using-coresh-locally">
<h1>Using CORESH locally<a class="headerlink" href="#using-coresh-locally" title="Link to this heading">#</a></h1>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Link to this heading">#</a></h2>
<p>This vignette demonstrates the core algorithm behind CORESH search
engine for querying public gene expression datasets based on a
user-provided gene signature (<a class="reference external" href="https://alserglab.wustl.edu/coresh/">https://alserglab.wustl.edu/coresh/</a>).
CORESH ranks the datasets based on the level of coregulation of
user-provided genes using a score inspired by Principal Component
Analysis, which can be applied to any gene expression matrix. Currently,
CORESH operates on a compendium of more than 40,000 mouse and 40,000
human gene expression datasets from the GEO database, including datasets
from both microarray and RNA-seq profiling.</p>
</section>
<section id="installing-dependencies">
<h2>Installing dependencies<a class="headerlink" href="#installing-dependencies" title="Link to this heading">#</a></h2>
<p>All of the required R dependencies can be installed by installing coresh
R package:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">remotes</span><span class="p">)</span>
<span class="nf">install_github</span><span class="p">(</span><span class="s">&quot;alserglab/coresh&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="getting-the-data">
<h2>Getting the data<a class="headerlink" href="#getting-the-data" title="Link to this heading">#</a></h2>
<p>The preprocessed compendium of GEO datasets can be downloaded from the
<a class="reference external" href="https://www.synapse.org/coresh">Synapse project syn66227307</a>. Notice,
that to download the files you will need to register at Synapse first.</p>
<p>After user registration, you can install the official Python client
(<a class="reference external" href="https://python-docs.synapse.org/en/stable/tutorials/installation/">https://python-docs.synapse.org/en/stable/tutorials/installation/</a>).
Install the client with pip using the command line:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>--upgrade<span class="w"> </span>synapseclient
</pre></div>
</div>
<p>Then log in to the client (you can use a personal access token, which
can be created at
<a class="reference external" href="https://accounts.synapse.org/authenticated/personalaccesstokens">https://accounts.synapse.org/authenticated/personalaccesstokens</a>):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>synapse<span class="w"> </span>config
</pre></div>
</div>
<p>Finally, download the files from the project</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>synapse<span class="w"> </span>get<span class="w"> </span>-r<span class="w"> </span>syn66227307
</pre></div>
</div>
<p>If everything finished correctly, a <code class="docutils literal notranslate"><span class="pre">preprocessed_chunks</span></code> folder should
appear.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tree<span class="w"> </span>-n<span class="w"> </span>preprocessed_chunks<span class="w"> </span><span class="p">|</span><span class="w"> </span>head
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## preprocessed_chunks
## ├── SYNAPSE_METADATA_MANIFEST.tsv
## ├── hsa
## │   ├── SYNAPSE_METADATA_MANIFEST.tsv
## │   ├── chunk_10_full_objects.qs2
## │   ├── chunk_11_full_objects.qs2
## │   ├── chunk_12_full_objects.qs2
## │   ├── chunk_13_full_objects.qs2
## │   ├── chunk_14_full_objects.qs2
## │   ├── chunk_15_full_objects.qs2
</pre></div>
</div>
</section>
<section id="running-coresh-analysis-locally">
<h2>Running CORESH analysis locally<a class="headerlink" href="#running-coresh-analysis-locally" title="Link to this heading">#</a></h2>
<p>Now we can switch to R and take a look at the data. In this vignette we
will be working with the human data.</p>
<p>The data is split into around hundred chunks:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">chunkPaths</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list.files</span><span class="p">(</span><span class="s">&quot;./preprocessed_chunks/hsa/&quot;</span><span class="p">,</span>
<span class="w">                            </span><span class="n">pattern</span><span class="o">=</span><span class="s">&quot;full_objects.qs2&quot;</span><span class="p">,</span>
<span class="w">                            </span><span class="n">full.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">chunkPaths</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## [1] 89
</pre></div>
</div>
<p>Each chunk consists from around 500 objects and can be loaded using qs2
packages:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">qs2</span><span class="p">)</span>
<span class="n">chunk</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">qs_read</span><span class="p">(</span><span class="n">chunkPaths</span><span class="p">[[</span><span class="m">1</span><span class="p">]])</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">chunk</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## [1] 500
</pre></div>
</div>
<p>Each object corresponds to a GEO dataset. Note that <code class="docutils literal notranslate"><span class="pre">E1024</span></code> attribute
contains a processed gene expression matrix: it is centered, potentially
reduced with a Principal Component Analysis, multiplied by 1024 and
rounded to the nearest integer.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">str</span><span class="p">(</span><span class="n">chunk</span><span class="p">[[</span><span class="m">1</span><span class="p">]])</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## List of 8
##  $ rownames  : int [1:10000] 9 12 14 16 18 19 20 21 22 23 ...
##  $ samples   : chr [1:10(1d)] &quot;GSM15785&quot; &quot;GSM15786&quot; &quot;GSM15787&quot; &quot;GSM15788&quot; ...
##  $ nsamples  : int 10
##  $ totalVar  : num 3096
##  $ E1024     : int [1:10000, 1:10] 28 51 120 -268 -368 -18 196 318 114 6 ...
##  $ gseId     : chr &quot;GSE1000&quot;
##  $ gplId     : chr &quot;GPL96&quot;
##  $ wordMatrix: num [1:15, 1:10] -0.158 -0.158 0.632 0.387 0.387 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:15] &quot;culture&quot; &quot;terminated&quot; &quot;sample&quot; &quot;acid&quot; ...
##   .. ..$ : chr [1:10] &quot;GSM15785&quot; &quot;GSM15786&quot; &quot;GSM15787&quot; &quot;GSM15788&quot; ...
</pre></div>
</div>
<p>As a test query gene set we will use “HALLMARK_HYPOXIA” gene set from
the MSigDB database:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">msigdbr</span><span class="p">)</span>
<span class="n">hallmarks</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">msigdbr</span><span class="p">(</span><span class="n">species</span><span class="o">=</span><span class="s">&quot;human&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">collection</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;H&quot;</span><span class="p">)</span>
<span class="n">query</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">hallmarks</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">    </span><span class="n">dplyr</span><span class="o">::</span><span class="nf">filter</span><span class="p">(</span><span class="n">gs_name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;HALLMARK_HYPOXIA&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span>
<span class="w">    </span><span class="n">dplyr</span><span class="o">::</span><span class="nf">pull</span><span class="p">(</span><span class="n">ncbi_gene</span><span class="p">)</span>

<span class="n">query</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">(</span><span class="n">query</span><span class="p">)</span><span class="w"> </span><span class="c1"># gene IDs are stored as integers</span>
<span class="nf">str</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>##  int [1:200] 57007 133 136 205 9590 226 229 230 272 51129 ...
</pre></div>
</div>
<p>Let’s define a function to match an object against the query gene set,
which will rely on <code class="docutils literal notranslate"><span class="pre">geseca</span></code> method from the <code class="docutils literal notranslate"><span class="pre">fgsea</span></code> package:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">fgsea</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span>
<span class="n">coreshMatch</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span><span class="w"> </span><span class="n">query</span><span class="p">,</span><span class="w"> </span><span class="n">calculatePvalues</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">E</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">obj</span><span class="o">$</span><span class="n">E1024</span><span class="o">/</span><span class="m">1024</span>
<span class="w">    </span><span class="n">queryIdxs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">na.omit</span><span class="p">(</span><span class="nf">match</span><span class="p">(</span><span class="n">query</span><span class="p">,</span><span class="w"> </span><span class="n">obj</span><span class="o">$</span><span class="n">rownames</span><span class="p">))</span>
<span class="w">    </span><span class="n">k</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">queryIdxs</span><span class="p">)</span>
<span class="w">        </span>
<span class="w">    </span><span class="n">curProfile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">colSums</span><span class="p">(</span><span class="n">E</span><span class="p">[</span><span class="n">queryIdxs</span><span class="p">,</span><span class="w"> </span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">])</span>
<span class="w">    </span><span class="n">queryVar</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">curProfile</span><span class="o">**</span><span class="m">2</span><span class="p">)</span>
<span class="w">    </span>
<span class="w">    </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">calculatePvalues</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">gesecaPval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fgsea</span><span class="o">:::</span><span class="nf">gesecaCpp</span><span class="p">(</span><span class="n">E</span><span class="p">,</span><span class="w"> </span><span class="n">queryVar</span><span class="p">,</span><span class="w"> </span><span class="n">k</span><span class="p">,</span><span class="w"> </span><span class="n">sampleSize</span><span class="o">=</span><span class="m">21</span><span class="p">,</span><span class="w"> </span><span class="n">seed</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">eps</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1e-300</span><span class="p">)[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="n">pval</span>
<span class="w">    </span><span class="p">}</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">gesecaPval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">NA</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span>
<span class="w">    </span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.table</span><span class="p">(</span>
<span class="w">        </span><span class="n">gse</span><span class="o">=</span><span class="n">obj</span><span class="o">$</span><span class="n">gseId</span><span class="p">,</span>
<span class="w">        </span><span class="n">gpl</span><span class="o">=</span><span class="n">obj</span><span class="o">$</span><span class="n">gplId</span><span class="p">,</span>
<span class="w">        </span><span class="n">pctVar</span><span class="o">=</span><span class="n">queryVar</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">k</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">obj</span><span class="o">$</span><span class="n">totalVar</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="p">,</span>
<span class="w">        </span><span class="n">pval</span><span class="o">=</span><span class="n">gesecaPval</span><span class="p">,</span>
<span class="w">        </span><span class="n">size</span><span class="o">=</span><span class="n">k</span>
<span class="w">    </span><span class="p">)</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Testing on the first dataset:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">coreshMatch</span><span class="p">(</span><span class="n">chunk</span><span class="p">[[</span><span class="m">1</span><span class="p">]],</span><span class="w"> </span><span class="n">query</span><span class="p">,</span><span class="w"> </span><span class="n">calculatePvalues</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>##        gse    gpl    pctVar         pval  size
##     &lt;char&gt; &lt;char&gt;     &lt;num&gt;        &lt;num&gt; &lt;int&gt;
## 1: GSE1000  GPL96 0.3102065 4.903416e-11   165
</pre></div>
</div>
<p>Before running it on the whole compendium, set up the parallel back-end
appropriate to your machine. The settings could be different from the
ones below.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">)</span>
<span class="n">bpparam</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BiocParallel</span><span class="o">::</span><span class="nf">MulticoreParam</span><span class="p">(</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">progressbar</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p>Without P-value calculation, the whole process should take 10-20
seconds, depending on your machine.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>##          gse      gpl   pctVar   pval  size
##       &lt;char&gt;   &lt;char&gt;    &lt;num&gt; &lt;lgcl&gt; &lt;int&gt;
## 1: GSE131379 GPL18573 7.452995     NA   147
## 2: GSE198308 GPL24676 7.033059     NA   150
## 3:  GSE59449  GPL6244 6.911020     NA   141
## 4: GSE239892 GPL18573 6.878768     NA   145
## 5: GSE196274 GPL23227 6.804668     NA   144
## 6: GSE179327 GPL11154 6.496734     NA   153
</pre></div>
</div>
<p>Ranking by p-value usually is a bit more specific, but takes about
couple of minutes to calculate.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">pvalRanking</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbindlist</span><span class="p">(</span><span class="nf">bplapply</span><span class="p">(</span><span class="n">chunkPaths</span><span class="p">,</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">chunkPath</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">chunk</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">qs_read</span><span class="p">(</span><span class="n">chunkPath</span><span class="p">)</span>
<span class="w">    </span><span class="n">chunkRanking</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbindlist</span><span class="p">(</span><span class="nf">lapply</span><span class="p">(</span><span class="n">chunk</span><span class="p">,</span><span class="w"> </span><span class="n">coreshMatch</span><span class="p">,</span><span class="w"> </span><span class="n">query</span><span class="o">=</span><span class="n">query</span><span class="p">,</span><span class="w"> </span><span class="n">calculatePvalue</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span>
<span class="w">    </span><span class="n">chunkRanking</span>
<span class="p">},</span><span class="w"> </span><span class="n">BPPARAM</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bpparam</span><span class="p">))</span>
<span class="n">pvalRanking</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">pvalRanking</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">pval</span><span class="p">)]</span>
<span class="nf">head</span><span class="p">(</span><span class="n">pvalRanking</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>##          gse      gpl   pctVar          pval  size
##       &lt;char&gt;   &lt;char&gt;    &lt;num&gt;         &lt;num&gt; &lt;int&gt;
## 1: GSE198308 GPL24676 7.033059 8.722967e-233   150
## 2: GSE145224 GPL20301 5.942319 2.008495e-229   159
## 3: GSE153557 GPL24676 5.627913 4.722340e-229   147
## 4:  GSE59449  GPL6244 6.911020 4.050704e-228   141
## 5: GSE153398 GPL11154 4.405773 2.327456e-166   144
## 6: GSE153291 GPL18573 4.320505 7.290046e-126   133
</pre></div>
</div>
<p>The current compendium snapshot does not provide metadata, but you can
get the information from GEO:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">geoUrl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sprintf</span><span class="p">(</span><span class="s">&quot;https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=%s&amp;targ=self&amp;form=text&amp;view=brief&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">pvalRanking</span><span class="o">$</span><span class="n">gse</span><span class="p">[</span><span class="m">1</span><span class="p">])</span>
<span class="nf">head</span><span class="p">(</span><span class="nf">readLines</span><span class="p">(</span><span class="n">geoUrl</span><span class="p">),</span><span class="w"> </span><span class="m">3</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## [1] &quot;^SERIES = GSE198308&quot;                                                                                              
## [2] &quot;!Series_title = RNA-seq of HeLa cells stably overexpressing SRSF6-GFP grown under normoxic and hypoxic conditions&quot;
## [3] &quot;!Series_geo_accession = GSE198308&quot;
</pre></div>
</div>
</section>
<section id="session-info">
<h2>Session info<a class="headerlink" href="#session-info" title="Link to this heading">#</a></h2>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">sessionInfo</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>## R version 4.5.0 (2025-04-11)
## Platform: aarch64-apple-darwin24.2.0
## Running under: macOS Sequoia 15.7
## 
## Matrix products: default
## BLAS:   /opt/homebrew/Cellar/openblas/0.3.29/lib/libopenblasp-r0.3.29.dylib 
## LAPACK: /opt/homebrew/Cellar/r/4.5.0/lib/R/lib/libRlapack.dylib;  LAPACK version 3.12.1
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: America/Chicago
## tzcode source: internal
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] BiocParallel_1.42.0 data.table_1.17.2   fgsea_1.35.6       
## [4] msigdbr_25.1.1      qs2_0.1.5          
## 
## loaded via a namespace (and not attached):
##  [1] Matrix_1.7-3          gtable_0.3.6          babelgene_22.9       
##  [4] dplyr_1.1.4           compiler_4.5.0        tidyselect_1.2.1     
##  [7] Rcpp_1.0.14           parallel_4.5.0        assertthat_0.2.1     
## [10] scales_1.4.0          yaml_2.3.10           fastmap_1.2.0        
## [13] lattice_0.22-6        ggplot2_3.5.2         R6_2.6.1             
## [16] generics_0.1.4        curl_6.2.2            knitr_1.50           
## [19] tibble_3.2.1          stringfish_0.17.0     pillar_1.10.2        
## [22] RColorBrewer_1.1-3    rlang_1.1.6           fastmatch_1.1-6      
## [25] xfun_0.52             RcppParallel_5.1.11-1 cli_3.6.5            
## [28] withr_3.0.2           magrittr_2.0.3        digest_0.6.37        
## [31] grid_4.5.0            rstudioapi_0.17.1     cowplot_1.1.3        
## [34] lifecycle_1.0.4       vctrs_0.6.5           evaluate_1.0.3       
## [37] glue_1.8.0            farver_2.1.2          codetools_0.2-20     
## [40] rmarkdown_2.29        tools_4.5.0           pkgconfig_2.0.3      
## [43] htmltools_0.5.8.1
</pre></div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./pages"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="hallmarks.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">CORESH Hallmark Gene Set Results</p>
      </div>
    </a>
    <a class="right-next"
       href="references.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">References</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#installing-dependencies">Installing dependencies</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#getting-the-data">Getting the data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#running-coresh-analysis-locally">Running CORESH analysis locally</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#session-info">Session info</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Vladimir Sukhov
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>